src/util/log.rs

```diff
@@ -17,153 +17,76 @@
-    loglevel.with(move |level| match ll {
-        LOG_TRACE..=LOG_FATAL => {
-            *level.borrow_mut() = ll;
-            Ok(())
-        }
-        _ => Err("Invalid log level".to_string()),
-    })
+lazy_static! {
+    pub static ref LOGGER: Logger = make_logger();
 }
 
-pub fn get_loglevel() -> u8 {
-    let mut res = 0;
-    loglevel.with(|lvl| {
-        res = *lvl.borrow();
-    });
+fn make_logger() -> Logger {
+    if env::var("BLOCKSTACK_LOG_JSON") == Ok("1".into()) {
+        let drain = Mutex::new(slog_json::Json::default(std::io::stderr())).map(slog::Fuse);
+        let filtered_drain = slog::LevelFilter::new(drain, get_loglevel()).fuse();
+        slog::Logger::root(filtered_drain, o!())
+    } else {
+        let decorator = slog_term::TermDecorator::new().build();
+        let drain = slog_term::CompactFormat::new(decorator).build().fuse();
+        let drain = slog_async::Async::new(drain).build().fuse();
+        let filtered_drain = slog::LevelFilter::new(drain, get_loglevel()).fuse();
+        slog::Logger::root(filtered_drain, o!())
+    }
+}
 
-    if env::var("BLOCKSTACK_DEBUG") == Ok("1".into()) && res > LOG_DEBUG {
-        set_loglevel(LOG_DEBUG).unwrap();
-        LOG_DEBUG
-    } else if env::var("BLOCKSTACK_TRACE") == Ok("1".into()) && res > LOG_TRACE {
-        set_loglevel(LOG_TRACE).unwrap();
-        LOG_TRACE
+fn get_loglevel() -> slog::Level {
+    if env::var("BLOCKSTACK_TRACE") == Ok("1".into()) {
+        slog::Level::Trace
+    } else if env::var("BLOCKSTACK_DEBUG") == Ok("1".into()) {
+        slog::Level::Debug
     } else {
-        res
+        slog::Level::Info
     }
 }
 
 #[macro_export]
 macro_rules! trace {
     ($($arg:tt)*) => ({
-        if ::util::log::get_loglevel() <= ::util::log::LOG_TRACE {
-            use std::time::SystemTime;
-            use std::thread;
-            let (ts_sec, ts_msec) = match SystemTime::now().duration_since(SystemTime::UNIX_EPOCH) {
-                Ok(n) => (n.as_secs(), n.subsec_nanos() / 1_000_000),
-                Err(_) => (0, 0)
-            };
-            eprintln!("TRACE [{}.{:03}] [{}:{}] [{:?}] {}", ts_sec, ts_msec, file!(), line!(), thread::current().id(), format!($($arg)*));
-        }
+        slog_trace!($crate::util::log::LOGGER, $($arg)*)
     })
 }
 
 #[macro_export]
-macro_rules! debug {
+macro_rules! error {
     ($($arg:tt)*) => ({
-        if ::util::log::get_loglevel() <= ::util::log::LOG_DEBUG {
-            use std::time::SystemTime;
-            use std::thread;
-            let (ts_sec, ts_msec) = match SystemTime::now().duration_since(SystemTime::UNIX_EPOCH) {
-                Ok(n) => (n.as_secs(), n.subsec_nanos() / 1_000_000),
-                Err(_) => (0, 0)
-            };
-            eprintln!("DEBUG [{}.{:03}] [{}:{}] [{:?}] {}", ts_sec, ts_msec, file!(), line!(), thread::current().id(), format!($($arg)*));
-        }
+        slog_error!($crate::util::log::LOGGER, $($arg)*)
```

Comment:
Instead of deleting this code (and the monitoring code that ties into it), can you preserve it, and make it so that the `slog` path only executes if `BLOCKSTACK_LOG_JSON` is set to `1`?  I'd prefer to have `slog` as a compile-time feature flag if possible.

Reply:
I think it's hard to make it a compile-time flag, because the macro has a different footprint: the slog macros accept key-value pairs (that's how you make the structured entries).

Reply:
Ah, right.